readme.md

```diff
@@ -64,7 +72,22 @@ The software changes required to support this consensus change differ significan
 
 ### Non-Inclusion of VM Number Format or Operation Descriptions
 
-Beyond dropping the unnecessary limit on VM number length, this proposal does not modify any other properties of the VM. Notably, the precise format and behavior of VM numbers across all VM operations – especially `OP_1ADD` (`0x8b`), `OP_1SUB` (`0x8c`), `OP_NEGATE` (`0x8f`), `OP_ABS` (`0x90`), `OP_NOT` (`0x91`), `OP_0NOTEQUAL` (`0x92`), `OP_ADD` (`0x93`), `OP_SUB` (`0x94`), `OP_MUL` (`0x95`), `OP_DIV` (`0x96`), `OP_MOD` (`0x97`), `OP_BOOLAND` (`0x9a`), `OP_BOOLOR` (`0x9b`), `OP_NUMEQUAL` (`0x9c`), `OP_NUMEQUALVERIFY` (`0x9d`), `OP_NUMNOTEQUAL` (`0x9e`), `OP_LESSTHAN` (`0x9f`), `OP_GREATERTHAN` (`0xa0`), `OP_LESSTHANOREQUAL` (`0xa1`), `OP_GREATERTHANOREQUAL` (`0xa2`), `OP_MIN` (`0xa3`), `OP_MAX` (`0xa4`), and `OP_WITHIN` (`0xa5`) – are part of network consensus and do not otherwise change as a result of this proposal. For the avoidance of doubt, see the [Tests & Benchmarks](#tests--benchmarks).
+As described in [Interaction with Existing Operations](#interaction-with-existing-operations), this proposal does not modify any other properties of the VM. Notably, the precise format and behavior of VM numbers across all VM operations is part of network consensus and does not change as a result of this proposal. For the avoidance of doubt, see the [Tests & Benchmarks](#tests--benchmarks).
+
+### Removal of Overflow Behavior
+
+As a practical consequence of removing the limit, numerical operation overflow is no longer a meaningful concept. A number that becomes too large would be too large due to the cost system established in [`CHIP: Targeted Virtual Machine Limits`](https://github.com/bitjson/bch-vm-limits), not due to a numerical boundary. These operations previously required overflow considerations: `OP_1ADD` (`0x8b`), `OP_1SUB` (`0x8c`), `OP_ADD` (`0x93`), `OP_SUB` (`0x94`), `OP_MUL` (`0x95`).
```

Comment:
Incorrect. **Output** of an arithmetic operation can overflow and fail because of it.

However, contracts will be safer because now any minimally encoded stack item will be **valid input** to arithmetic opcodes.

Reply:
i.e. we can't anymore overflow on input, but can still overflow on output

Reply:
Ok... I haven't seen that specified anywhere. I have only seen "remove limit". What's the reasoning for leaving overflow?

Reply:
How do you not leave it? When you say "not having overflow" to me it would imply circumventing the stack size limit and allowing >10,000 stack item if it's a result of arithmetic op which is obv. not what's happening. There's always a limit and hitting a limit is normally called overflow.
The effective output overflow limit is now `BigInt(2).pow(MAX_ELEM_SIZE * 8 - 1) - 1 == MAX_SCRIPTNUM`.
Because BigInt library supports infinite precision (bounded by hardware), we can actually calculate an overflowing result and *then* fail it for being over the limit, but it still must fail. From PoV of ScriptVM - it *is* an overflow, it's weird not to call it that.

Reply:
"Did I just miss the specification?" - it was not explicitly specified, but implicitly the limit is now 2^(10,000 * 8 - 1) - 1.

It's better to have state it explicitly to prevent misunderstandings so I have these lines in my PR:

- https://github.com/bitjson/bch-bigint/pull/4/files#diff-5a831ea67cf5cf8703b0de46901ab25bd191f56b320053be9332d9a3b0d01d15R48
- https://github.com/bitjson/bch-bigint/pull/4/files#diff-bc6661da34ecae62fbe724bb93fd69b91a7f81143f2683a81163231de7e3b545R71

Reply:
Hmm. I'm still not completely clear on it.

There is a difference between "overflow is effectively enforced" and "overflow is explicitly enforced". In current behavior, they may be equivalent (although it needs to be specified exactly which one is the case) but the situation will be different in some future upgrade where the difference and intent becomes relevant. Not to mention closing the gap for any potential bugs resulting from misunderstanding.

Which one is it?

a) the byte size enforced, resulting in an effective maxint
b) an explicit maxint is enforced
c) something else

If a) then what we wrote is strictly and precisely correct, isn't it?

If b) then it needs to be explicitly clarified that overflow is still directly enforced.

Reply:
a)

>If a) then what we wrote is strictly and precisely correct, isn't it?

No because "numerical operation overflow is no longer a meaningful concept. A number that becomes too large would be too large due to the cost system established" is plain wrong still.

- it *is* a meaningful concept, even if effective result of the stack item size limit
- it's not because of "cost system" (cost accumulating throughout execution of input script) but because of the specific limit on max. stack item size - which is a static constant

now... implementations can opt for b) even though the specification doesn't require it, and BCHN did, and my property tests check their particular implementation in the [assumptions test](https://gitlab.com/cculianu/bitcoin-cash-node/-/blob/wip_bca_script_big_int/src/test/bigint_script_property_tests.cpp#L380):

```

BOOST_AUTO_TEST_CASE(assumptions) {
    CHECK_MESSAGE(MAX_SCRIPTNUM == -MIN_SCRIPTNUM, "Assumption failed: MAX_SCRIPTNUM == -MIN_SCRIPTNUM");
    CHECK_MESSAGE(BigInt(2).pow(MAX_ELEM_SIZE * 8 - 1) - 1 == MAX_SCRIPTNUM, "Assumption failed: BigInt(2).pow(MAX_ELEM_SIZE * 8 - 1) - 1 == MAX_SCRIPTNUM");
}
```

Reply:
also... if *output* size is consistently enforced, then there's no need to specifically enforce *input* size because it is simply not possible to have an item on stack that would exceed the limit - how would it get produced to begin with?

Reply:
Ok so what we wrote is correct and precise if we change it from "cost system" to stack item size.

The specification needs to make that clear. Where you think there is no room for a gap between A and B, often hides a dragon.

Reply:
parameters are outputs of push opcodes, and those have the stack item size limit, too

Reply:
Thanks for discussion. @bitjson could I ask for a clarification of intent? I'll revise this PR after that.